addons/website/models/ir_http.py

```diff
@@ -83,6 +83,32 @@ def _slug_matching(cls, adapter, endpoint, **kw):
         qs = request.httprequest.query_string.decode('utf-8')
         return adapter.build(endpoint, kw) + (qs and '?%s' % qs or '')
 
+    @classmethod
+    def _url_for(cls, url_from: str, lang_code: str | None = None) -> str:
+        ''' Return the url with the rewriting applied.
+            Nothing will be done for absolute URL, invalid URL, or short URL from 1 char.
+
+            :param url_from: The URL to convert.
+            :param lang_code: Must be the lang `code`. It could also be something
+                              else, such as `'[lang]'` (used for url_return).
+        '''
+        path, _, qs = (url_from or '').partition('?')
+        if (
+            path
+            # don't try to match route if we know that no rewrite has been loaded.
+            and request.env['ir.http']._rewrite_len(request.website_routing)
+            and (
+                len(path) > 1
+                and path.startswith('/')
+                and '/static/' not in path
+                and not path.startswith('/web/')
+            )
+        ):
+            url_from, _ = request.env['ir.http'].url_rewrite(path)
+            url_from = url_from if not qs else url_from + '?%s' % qs
+
+        return super()._url_for(url_from, lang_code)
```

Comment:
Commit message `[IMP] base: Make moving _url_for on ir.http compositional ` does not help me: what is this about? Did we have code in http_routing that just was supposed to be website only from the start? :shrug:

Reply:
The part of the code above was only valid when website was installed because of the `getattr` as well as the `_rewrite_len` method which always returned `0`.